### PR TITLE
feat!(Streams): Support propagating Unfolds; remove StreamResult

### PR DESCRIPTION
**Pull request by bartelink**
Enhances the pipeline to support propagation of `Unfold` events through the pipeline:
- `EquinoxSystemTextJsonParser.eventsAndUnfoldsWhereStream` extracts the unfolds in addition to the events
- `StreamSpan.merge` ensures we only retain the most recent ones as inputs are coalesced
- Handlers are presented any unfolds at the end of the `span` (assuming they fit inside the limit)
- Unfolds will always have an `IsUnfold = true` and `Index` = _version_, where version is _the Index of the last 'real' event +1_
- For Sync, EventStoreDbSink and CosmosStoreSink, the `maxBytes` limit will include/exclude either all or none of the unfolds
- Retries pick up the most recent unfold

As before, given a `Propulsion.Sink.Event[]`, you compute:
- the index of the first event using `Propulsion.Sink.Event.index`
- the index of the next event using `Propulsion.Sink.Event.nextIndex` (even if there are unfolds at the end of the sequence)

Note that the following assumptions no longer hold if you have unfolds in an `Event[]`:
- `nextIndex` <-> `index events + events.Length` (renamed to `next` as intentional breaking change)
- actual events (not unfolds) count <-> `events.Length`

----

What is the use case? If you're running a ChangeFeedProcessor from an Equinox.CosmosStore, the changefeed includes:
- snapshots (which you could hydrate the state from directly without reading the store using Equinox)
- `RollingState` state as held in the tip

Being able to route the Unfolds to the Handler alongside normal events enables:
- complete copying of the content of a store in a single traversal via (`propulsion sync cosmos` or `propulsion sync file cosmos`)
- being able to run a sync job (or generic event handler) on a rolling state category
- being able to avoid Equinox reads of the source streams in the bulk of cases (as you can hydrate from the Unfolds instead of loading the state) using features in https://github.com/jet/equinox/pull/450

### CODE REVIEW

File: src/Propulsion.CosmosStore/CosmosStorePruner.fs [hunk 1]
```diff
@@ -29,7 +29,7 @@ module Pruner =
         let res = if deleted = 0 && deferred = 0 then Nop span.Length else Ok (deleted, deferred)
         // For case where we discover events have already been deleted beyond our requested position, signal to reader to drop events
         let writePos = max trimmedPos (untilIndex + 1L)
-        return struct (writePos, res) }
+        return struct (res, writePos) }
```
Comment (bartelink) [inline src/Propulsion.CosmosStore/CosmosStorePruner.fs:32]: for alignment with `res, events` with equinox result tuples

File: src/Propulsion.CosmosStore/CosmosStorePruner.fs [hunk 2]
```diff
@@ -75,8 +75,8 @@ type CosmosStorePruner =
 #endif
             let interpret _stream span =
                 let metrics = StreamSpan.metrics Event.storedSize span
-                struct (metrics, span)
-            Dispatcher.Concurrent<_, _, _, _>.Create(maxConcurrentStreams, interpret, Pruner.handle pruneUntil, (fun _ r -> r))
+                struct (span, metrics)
```
Comment (bartelink) [inline src/Propulsion.CosmosStore/CosmosStorePruner.fs:78]: any extra stuff layered on goes after the primary data in the tuple - no idea how/why I ended up having it backwards, but these signatures are undergoing change as part of putting the pos after the result anyway

File: src/Propulsion.CosmosStore/CosmosStoreSink.fs [hunk 3]
```diff
@@ -49,38 +47,42 @@ module Internal =
...
+                log.Information("Waiting   {stream} missing {gap} events before {pos}", stream, gap, pos)
+            | stream, Error (exn, _) ->
                 let level = if malformed then LogEventLevel.Warning else Events.LogEventLevel.Information
                 log.Write(level, exn, "Writing   {stream} failed, retrying", stream)
 
         let write (log: ILogger) (ctx: EventsContext) stream (span: Event[]) ct = task {
-            log.Debug("Writing {s}@{i}x{n}", stream, span[0].Index, span.Length)
+            let i = StreamSpan.index span
+            let n = StreamSpan.nextIndex span
+            span |> Seq.iter (fun x -> if x.IsUnfold then invalidOp "CosmosStore3 does not [yet] support ingesting unfolds")
```
Comment (bartelink) [inline src/Propulsion.CosmosStore/CosmosStoreSink.fs:69]: prep for #263

File: src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs [hunk 4]
```diff
@@ -33,36 +32,55 @@ module EquinoxSystemTextJsonParser =
...
             let inline len s = if isNull s then 0 else String.length s
-            FsCodec.Core.TimelineEvent.Create(batch.i + int64 offset, x.c, d, m, timestamp = x.t,
+            FsCodec.Core.TimelineEvent.Create(i, x.c, d, m, timestamp = x.t,
                                               size = x.c.Length + d.Length + m.Length + len x.correlationId + len x.causationId + 80,
-                                              correlationId = x.correlationId, causationId = x.causationId))
-
-    /// Attempts to parse a Document/Item from the Store
+                                              correlationId = x.correlationId, causationId = x.causationId, isUnfold = isUnfold)
+        let events = batch.e |> Seq.mapi (fun offset -> gen false (batch.i + int64 offset))
+        // an Unfold won't have a corr/cause id, but that's OK - can't use Tip type as don't want to expand compressed form etc
```
Comment (nordfjord) [inline src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs:48]: v5 wishlist: drop correlation/causation as first class things 😇

File: src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs [hunk 5]
```diff
@@ -33,36 +32,55 @@ module EquinoxSystemTextJsonParser =
...
+        match u |> ValueOption.map (fun u -> u.Cast<Equinox.CosmosStore.Core.Event[]>()) with
+        | ValueNone | ValueSome null | ValueSome [||] -> events
+        | ValueSome unfolds -> seq {
+            yield! events
+            for x in unfolds do
+                gen true batch.n x }
+    let inline tryEnumStreamEvents_ withUnfolds streamFilter jsonDocument: seq<StreamEvent> voption =
+        tryParseEquinoxBatchOrTip streamFilter jsonDocument
+        |> ValueOption.map (fun struct (s, xs, u) -> enumEquinoxCosmosBatchOrTip (if withUnfolds then u else ValueNone) xs |> Seq.map (fun x -> s, x))
```
Comment (nordfjord) [inline src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs:57]: this'd read better as a plain 'ole match

File: src/Propulsion.Feed/PeriodicSource.fs [hunk 6]
```diff
@@ -38,7 +38,7 @@ module private TimelineEvent =
         fun (i, x: FsCodec.IEventData<_>, context: obj) ->
             if i > DateTimeOffsetPosition.factor then invalidArg (nameof i) $"Index may not exceed %d{DateTimeOffsetPosition.factor}"
             FsCodec.Core.TimelineEvent.Create(
-                baseIndex + i, x.EventType, x.Data, x.Meta, x.EventId, x.CorrelationId, x.CausationId, x.Timestamp, isUnfold = true, context = context)
```
Comment (nordfjord) [inline src/Propulsion.Feed/PeriodicSource.fs:41]: whoopsie?

File: src/Propulsion/Sinks.fs [hunk 7]
```diff
@@ -17,36 +17,9 @@ type Codec<'E> = FsCodec.IEventCodec<'E, EventBody, unit>
...
-    let nextIndex: Event[] -> int64 = Streams.StreamSpan.ver
+    let nextIndex: Event[] -> int64 = Streams.StreamSpan.nextIndex
     /// The Index of the first event as supplied to this handler
-    let index: Event[] -> int64 = Streams.StreamSpan.idx
-
-/// Represents progress attained during the processing of the supplied Events for a given <c>StreamName</c>.
-/// This will be reflected in adjustments to the Write Position for the stream in question.
-/// Incoming <c>StreamEvent</c>s with <c>Index</c>es prior to the Write Position implied by the result are proactively
-/// dropped from incoming buffers, yielding increased throughput due to reduction of redundant processing.
-type StreamResult =
```
Comment (nordfjord) [inline src/Propulsion/Sinks.fs:28]: Moving this is a breaking change

File: src/Propulsion/Streams.fs [hunk 8]
```diff
@@ -84,57 +84,78 @@ module StreamSpan =
...
-                if span <> null then
-                    match dropBeforeIndex min span with
-                    | null -> ()
-                    | trimmed when trimmed.Length = 0 -> invalidOp "Cant add empty"
-                    | trimmed -> trimmed |]
+        let candidates = [| for span in spans do
+                                if span <> null then
+                                    match dropBeforeIndex min span with
+                                    | [||] -> ()
+                                    | xs -> xs |]
```
Comment (nordfjord) [inline src/Propulsion/Streams.fs:127]: was better before the added indentation 😓

**Review by nordfjord — APPROVED** [review]
Not picking up on anything too bad, but this would be a new major version since the user facing API seems to be changing

File: src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs (hunk 4 above)
Comment (bartelink) [reply src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs:48]: yes, it's a misfeature that's one of the few white elephants in Equinox, but it does happen to be surfaced as top level indexable value in Equinox.CosmosStore
AFAIK nobody is actually leaning on it, but I could be wrong, so there'd conceptually have to be a placeholder task to provide a way to post process IEventData -> Event mapping in both directions.

File: src/Propulsion.Feed/PeriodicSource.fs (hunk 6 above)
Comment (bartelink) [reply src/Propulsion.Feed/PeriodicSource.fs:41]: no, it's intentional - this first/only use of the flag was to convey that these were ephemeral events, not backed by a store
but nothing leans on that
the change in this PR as a whole is that IsUnfold events now get dropped and replaced whenever a new set comes through (similar to how cosmos changefeed will deliver only the latest state)
But that would be bad for periodic source, which generates a set of sequntial events to represent the full state of the input at a point in time (and then re-runs the search with a new prefix when the period resets)

File: src/Propulsion/Sinks.fs (hunk 7 above)
Comment (bartelink) [reply src/Propulsion/Sinks.fs:28]: it's only a RC 😢

File: src/Propulsion/Streams.fs (hunk 8 above)
Comment (bartelink) [reply src/Propulsion/Streams.fs:127]: it was an improvement at one point in the refactoring cycle, I promise, but agreed...

**Comment by bartelink** [comment]
> but this would be a new major version since the user facing API seems to be changing

You are not wrong of course... But I have some buts:

This handler signature was only introduced in the V3 timeframe, and I believe it to have been a misdesign - you want calculating the next version (or having it flow from what the sycing is going into) be foremost in people's minds

For a V4, I think Propulsion.Feed and Propulsion should be merged as lots of modules depend on Propulsion.Feed and it in turn depends on Propulsion

What if I promised to mint a V4 as soon as Equinox.CosmosStore 4.1.0 is final ? (and kill Propulsion.CosmosStore3 the minute that's done)

Other things on the list that I'd like to do at some point (which don't change the V4 API but simplify the whole deal):
- align Propulsion.Kafka's Source mechanism with how all the other sources work (it has lots of ugly parallel implementations that are pretty close to what Propulsion.Feed generalizes for everything else
- other stuff in https://github.com/jet/propulsion/issues/116 like folding in the EventStore striped reading stuff into Propulsion.EventStoreDb

File: src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs (hunk 5 above)
Comment (bartelink) [reply src/Propulsion.CosmosStore/EquinoxSystemTextJsonParser.fs:57]: It didn't but I left it!

File: src/Propulsion/Sinks.fs (hunk 7 above)
Comment (bartelink) [reply src/Propulsion/Sinks.fs:28]: Further, the original transition to this name/structure happened in V3 and I consider it a bug - my thinking was that the name was something to hang documentation off.

On reflection, confronting people with an `int64` makes a lot more sense - you have to think, and you don't have to explain the `OverrideNextIndex` name

**Pull request merged**